### PR TITLE
Add until filter to docker image ls

### LINKED ISSUE #46576: Add until filter to docker image ls (classic image backend)
### Description

Currently, `docker image ls` only accepts `since` and `before` filters, which are based on image ids. As seen [here](https://github.com/docker/cli/issues/3962) and [here](https://stackoverflow.com/questions/24659300/how-to-use-docker-images-filter/75095332#75095332), this feature would be useful to people who want to list out images that were created before a particular time. For example, one should be to do: `docker image ls --filter "until=6m"`

### PR DESCRIPTION
**Pull request by RachitSharma2001**
- Fixes https://github.com/moby/moby/issues/46576

**- What I did**
Added the ability to filter the results of `docker image ls` or `docker images` through timestamps by adding the `until` filter.

**- How I did it**
Modified the `acceptedImageFilterTags` map in `daemon/containerd/image_list.go` to include an `until` key. Then I modified `setupFilters` function in the same file to process the value assigned to the until key

**- How to verify it**
I added an integration test function `TestImagesFilterUntil`. I did not see any unit tests that test the `setupFilters` function directly, which is why I decided to add this new test instead. Let me know if there is a simpler or better way to test this. 

**- Description for the changelog**

`docker image ls` now allows filtering by timestamp

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/images/image_list.go
Comment (corhere) [inline daemon/images/image_list.go:91]: The callback function will be called for each `until` condition in the filter. There may be more than one `until` condition in the request. When there are multiple `until` conditions, which timestamp is used when filtering images? [It can't be last-one-wins.](https://github.com/moby/moby/pull/44503)

File: daemon/images/image_list.go
Comment (corhere) [inline daemon/images/image_list.go:120]: The `untilFilter` variable has identical semantics to `beforeFilter`: select images not created before that timestamp. Therefore the `untilFilter` variable is not needed. Update the `beforeFilter` variable instead.

**Review by corhere — CHANGES_REQUESTED** [review]
Thank you for contributing! It will be very helpful to have the classic image service support the same set of filter criteria as the containerd image service. Unfortunately a bit more work is needed to properly handle some edge cases.

File: daemon/images/image_list.go
Comment (RachitSharma2001) [reply daemon/images/image_list.go:91]: That is a good point. I will make a modification that will, out of all the timestamps given by before and until filters, choose the oldest timestamp. This way we can get rid of the `untilFilter` variable as well. I should be able to implement and push this modification in next 3 - 4 days.

**Comment by RachitSharma2001** [comment]
I have pushed a solution that addresses the issue of multiple filters. If multiple `until` and `before` filters are passed in, then only the one with the oldest timestamp will be used for filtering. The solution I pushed also eliminates the `untilFilter` variable, and updates the integration test for this feature to test for multiple filters. Let me know if there are any issues.

**Review by corhere — APPROVED** [review]
LGTM, thanks!

File: integration/image/list_test.go
Comment (corhere) [inline integration/image/list_test.go:75]: The linter is complaining that `types.ContainerCommitOptions` is deprecated.
```suggestion
		id, err := container.CommitOptions(ctx, ctr, types.ContainerCommitOptions{Reference: ctrName})
```

**Comment by thaJeztah** [comment]
Overall this looks good (thanks!). Looks like we also need to update the swagger file as well (but I see there's a "docs/revisit" label added)
https://github.com/moby/moby/blob/b4a08b3b7cc9cbffee008ce0d4d472d300071477/api/swagger.yaml#L8061-L8073

And we should add a bullet to the API version-history; https://github.com/moby/moby/blob/42a51cb2859d1b73b18a5e80c20865cc44581e9d/docs/api/version-history.md

As we (currently) don't gate filters by API version, we should mention that (we have some older entries that are also unversioned, so we can take the same text for that), e.g.;

> `GET /images/json` now accepts an `until` filter. This change is not versioned, and affects all API versions if the daemon has this patch.

(perhaps it should be more descriptive what it does / what kind of values it accepts, but I'd be ok with improving later)

**Comment by RachitSharma2001** [comment]
Thank you for the suggestions. I have committed the requested changes, including updating the documentation at both `moby/api/swagger.yaml` and `docs/api/version-history.md`. Let me know if there is anything else I should add.

File: api/swagger.yaml
Comment (thaJeztah) [inline api/swagger.yaml:8048]: Looks like `type: "string"` was dropped here; was that intentional or an accidental change?

Comment (RachitSharma2001) [reply api/swagger.yaml]: Sorry for that, it was accidental. I have pushed a fix.

File: integration/image/list_test.go
Comment (corhere) [inline integration/image/list_test.go:75]: Sorry about that, I completely botched my earlier suggestion. Let's try this again:
```suggestion
		id, err := client.ContainerCommit(ctx, ctr, container.CommitOptions{Reference: ctrName})
```

File: docs/api/version-history.md
Comment (thaJeztah) [inline docs/api/version-history.md:20]: We should probably wrap this to ~80 chars to match the formatting we use for markdown files.

**Comment by thaJeztah** [comment]
Let me ask if someone is able to do a second review; before merging, we should squash (some / all?) the commits in this PR.

**Comment by RachitSharma2001** [comment]
Sounds good. I have added wrapping of the text in `version-history`, and have squashed all the commits. Let me know if there is anything else that needs to be done.

File: integration/image/list_test.go
Comment (corhere) [inline integration/image/list_test.go:87]: The `Created` timestamp of an image is the timestamp of when the image was created. Images created by committing a container are no exception. The `Created` timestamp of the source container is irrelevant information. (Think about it. The container could have been created days, months, or years in the past. And containers are mutable: images of the same container committed at different times could be vastly different in content.) **The `olderUntil` and `laterUntil` timestamps need to be (derived from) the `Created` timestamps of the images themselves to ensure the test is not flaky.** As there is no correlation between the image timestamps and any other timestamp aside from the order of operations of the test itself, using any other timestamps is not much better than just calling `time.Now()` directly at the right points in the test.

For the same reason, only one source container is needed, just like in `TestImagesFilterBeforeSince`.

You only need the image ID to query its `Created` timestamp. Since the committed images' IDs are assigned to the `imgs` slice, the code to fetch the `olderUntil` and `laterUntil` values [can be moved to below the loop.](https://en.wikipedia.org/wiki/Loop-switch_sequence)

**Review by corhere — CHANGES_REQUESTED** [review]
My apologies for not spotting the logic bug in the test earlier.

**Review by corhere — APPROVED** [review]
LGTM! You understood the assignment perfectly.

**Comment by thaJeztah** [comment]
CI failure is unrelated; looks like something broke by https://github.com/moby/moby/pull/45905

```
integration/networking/bridge_test.go:98:30: not enough arguments in call to container.WithMacAddress
	have (string)
	want (string, string)
integration/networking/bridge_test.go:111:30: not enough arguments in call to container.WithMacAddress
	have (string)
	want (string, string)
make: *** [Makefile:212: test-integration] Error 1
Error: Process completed with exit code 2.
```

**Comment by thaJeztah** [comment]
Opened a PR to fix that failure; https://github.com/moby/moby/pull/46783

File: daemon/images/image_list.go
Comment (thaJeztah) [inline daemon/images/image_list.go:74]: I was initially looking if we could skip the handling in the `beforeFilter.IsZero()` case, looks like we can't because of the `beforeFilter.After(timestamp)`. We can still optimise in future if we find an alternative.

**Review by thaJeztah — APPROVED** [review]
LGTM

I did a quick rebase to make CI pass now that the fix was merged; we can merge this PR once CI finishes

Thank you!

**Pull request merged**
